brave/src/main/java/brave/Tag.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave;
+
+import brave.handler.FinishedSpanHandler;
+import brave.handler.MutableSpan;
+import brave.internal.Nullable;
+import brave.internal.Platform;
+import brave.propagation.TraceContext;
+
+import static brave.internal.Throwables.propagateIfFatal;
+
+/**
+ * This is a centralized type to parse a tag into any variant of a span. This also avoids the
+ * clutter of checking null or guarding on exceptions.
+ *
+ * Here's an example of a potentially expensive tag:
+ * <pre>{@code
+ * SUMMARY_TAG = new Tag<Summarizer>("summary") {
+ *   @Override protected String parseValue(Summarizer input, TraceContext context) {
+ *     return input.computeSummary();
+ *   }
+ * }
+ * SUMMARY_TAG.tag(span);
+ * }</pre>
+ *
+ * @see Tags
+ * @see SpanCustomizer#tag(String, String)
+ * @see ScopedSpan#tag(String, String)
+ * @see MutableSpan#tag(String, String)
+ * @since 5.11
+ */
+public abstract class Tag<I> {
+  public final String key() {
+    return key;
+  }
+
+  /**
+   * Override to change what data from the input are parsed into the span modeling it. Any
+   * exceptions will be logged and ignored.
+   *
+   * @return possibly nullable result. Note: empty string is a valid tag value!
+   * @since 5.11
+   */
+  @Nullable protected abstract String parseValue(I input, @Nullable TraceContext context);
+
+  /**
+   * {@linkplain ScopedSpan#tag(String, String) Tags} the value parsed from the {@code input} when
```

Comment:
Maybe can leave off these docs at `Tags the value parsed from the {@code input}.`. The non-null is really a job of parseValue, and there's no way to tag a no-op span anyways, the key point is we don't waste time parsing when it's no-op but that's an implementation detail.

Reply:
cool